api.lua

```diff
@@ -1,61 +1,121 @@
 hopper.containers = {}
 hopper.groups = {}
 hopper.neighbors = {}
+hopper.extra = {}
+
+local function parse_group(target_node)
+	local number
+	local identifier
+
+	local equals_index = string.find(target_node, "=")
+	if equals_index ~= nil then
+		identifier = string.sub(target_node, 7, equals_index-1)
+		-- it's possible that the string was of the form "group:blah = 1", in which case we want to trim spaces off the end of the group identifier
+		local space_index = string.find(identifier, " ")
+		if space_index ~= nil then
+			identifier = string.sub(identifier, 1, space_index-1)
+		end
+		number = tonumber(string.sub(target_node, equals_index+1, -1))
+	else
+		identifier = string.sub(target_node, 7, -1)
+		number = "all" -- special value to indicate no number was provided
+	end
+
+	return identifier, number
+end
+
+local function is_already_in_neighbors(neighbor_node)
+	for _, value in pairs(hopper.neighbors) do
+		if value == neighbor_node then
+			return true
+		end
+	end
+	return false
+end
 
 -- global function to add new containers
 function hopper:add_container(list)
 	for _, entry in pairs(list) do
-	
+
 		local target_node = entry[2]
 		local neighbor_node
-		
+
 		if string.sub(target_node, 1, 6) == "group:" then
-			local group_identifier, group_number
-			local equals_index = string.find(target_node, "=")
-			if equals_index ~= nil then
-				group_identifier = string.sub(target_node, 7, equals_index-1)
-				-- it's possible that the string was of the form "group:blah = 1", in which case we want to trim spaces off the end of the group identifier
-				local space_index = string.find(group_identifier, " ") 
-				if space_index ~= nil then
-					group_identifier = string.sub(group_identifier, 1, space_index-1)
-				end
-				group_number = tonumber(string.sub(target_node, equals_index+1, -1))
-			else
-				group_identifier = string.sub(target_node, 7, -1)
-				group_number = "all" -- special value to indicate no number was provided
-			end
-			
+			local group_identifier, group_number = parse_group(target_node)
+
 			local group_info = hopper.groups[group_identifier]
 			if group_info == nil then
 				group_info = {}
 			end
 			if group_info[group_number] == nil then
 				group_info[group_number] = {}
 			end
-			group_info[group_number][entry[1]] = entry[3]
+			if group_info[group_number][entry[1]] == nil then
+				group_info[group_number][entry[1]] = {}
+			end
```

Comment:
The following code is very repetitive. Please use an additional variable (references the table):
```Lua
local group = group_info[group_number]
```

Reply:
Applied.